### PR TITLE
Standard Solc Compile (Multiversion)

### LINKED ISSUE #2413: Use standard JSON compile inputs for solidity
Memorial for #2051 

Currently nucypher uses multi-version compiler inputs, which causes semi-incompatibility with third party systems and libraries like brownie, etherscan, static analyzers, etc.

### PR DESCRIPTION
**Pull request by vzotova**
Continued work on #2051

- Deprecates `SolidityCompiler` - Rework for a functional approach
- Introduces `compile` package
- Updates solidity compilation from the legacy to the standard compile method.
- Accommodate Standard JSON input configuration values
- ~Hypothesis [testing for devdoc regex full-matches](https://github.com/nucypher/nucypher/pull/2051/files?file-filters%5B%5D=.lock&file-filters%5B%5D=.py&file-filters%5B%5D=.rst&file-filters%5B%5D=.sol&file-filters%5B%5D=.txt&file-filters%5B%5D=No+extension&file-filters%5B%5D=dotfile&hide-deleted-files=true#diff-2befbeecc854fdf7de827acbcc9742a0)~
- Fixes #2413

### CODE REVIEW

File: nucypher/types.py [hunk 1]
```diff
@@ -30,7 +31,7 @@
 
 
 class ContractParams(Tuple):
-    pass
+    pass  # TODO
```
Comment (derekpierre) [inline nucypher/types.py:34]: Follow-up issue?

File: nucypher/blockchain/eth/interfaces.py [hunk 2]
```diff
@@ -603,6 +616,7 @@ def sign_and_broadcast_transaction(self,
         # Broadcast
         #
 
+        time.sleep(1)  # TODO: Avoids race condition from invalid/resused nonce.
```
Comment (derekpierre) [inline nucypher/blockchain/eth/interfaces.py:619]: Follow-up issue?

File: tests/acceptance/blockchain/deployers/test_token_deployer.py [hunk 3]
```diff
@@ -25,7 +25,7 @@ def test_token_deployer_and_agent(testerchain, deployment_progress, test_registr
 
     origin = testerchain.etherbase_account
 
-    # Trying to get token from blockchain before it's been published fails
+    # Trying to get token from blockchain it's been published fails
```
Comment (derekpierre) [inline tests/acceptance/blockchain/deployers/test_token_deployer.py:28]:
```suggestion
    # Trying to get token from blockchain before it's been published should fail
```

File: tests/mock/interfaces.py [hunk 4]
```diff
@@ -73,14 +73,18 @@ def fetch_latest_publication(self) -> Union[str, bytes]:
...
 
 class MockEthereumClient(EthereumClient):
 
     def __init__(self, w3):
-        super().__init__(w3, None, None, None, None)
+        super().__init__(w3=w3, node_technology=None, version=None, platform=None, backend=None)
+
+    def connect(self, *args, **kwargs) -> bool:
+        if 'compile_now' in kwargs:
+            raise ValueError("Mock testerchain cannot handle solidity soure compilation.")
```
Comment (derekpierre) [inline tests/mock/interfaces.py:89]:
```suggestion
            raise ValueError("Mock testerchain cannot handle solidity source compilation.")
```

**Review by derekpierre — APPROVED** [review]
🎸 - some minor suggestions/comments.

File: nucypher/blockchain/eth/interfaces.py (hunk 2 above)
Comment (vzotova) [reply nucypher/blockchain/eth/interfaces.py:619]: @KPrasch is it new issue?

File: nucypher/types.py (hunk 1 above)
Comment (vzotova) [reply nucypher/types.py:34]: @KPrasch

Comment (KPrasch) [reply nucypher/types.py:34]: For now, I suggest simply removing this type.

Comment (vzotova) [reply nucypher/types.py]: ✔️

File: nucypher/blockchain/eth/interfaces.py (hunk 2 above)
Comment (vzotova) [reply nucypher/blockchain/eth/interfaces.py]: removed this line

**Review by KPrasch — APPROVED** [review]
As a partial author of this PR it's difficult to give a complete review.  Overall code quality is good, and I am looking forward to being able to use standard JSON compile for etherscan verification :+1:

File: tests/utils/blockchain.py [hunk 5]
```diff
@@ -72,56 +74,50 @@ class TesterBlockchain(BlockchainDeployerInterface):
...
+    SOURCES: List[SourceBundle] = [
+        SourceBundle(base_path=SOLIDITY_SOURCE_ROOT,
+                     other_paths=(TEST_SOLIDITY_SOURCE_ROOT,))
+    ]
 
-    GAS_STRATEGIES = {**BlockchainDeployerInterface.GAS_STRATEGIES,
-                      'free': free_gas_price_strategy}
+    # Web3
+    GAS_STRATEGIES = {**BlockchainDeployerInterface.GAS_STRATEGIES, 'free': free_gas_price_strategy}
+    PROVIDER_URI = 'tester://pyevm'
```
Comment (cygnusv) [inline tests/utils/blockchain.py:87]: Use `tests.constants.PYEVM_DEV_URI`

**Review by cygnusv — APPROVED** [review]
Great work @vzotova and @KPrasch! I have some minor comments here and there, but in general looking very good.

File: tests/markers.py [hunk 6]
```diff
@@ -16,8 +16,8 @@
 """
 
 import os
-import pytest
 
+import pytest
```
[the remark this answers is not in the document]
Comment (vzotova) [reply tests/markers.py]: Could be some errors from rebasing, anyway I removed this commit

File: nucypher/blockchain/eth/interfaces.py [hunk 7]
```diff
@@ -15,6 +15,13 @@
 along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
 """
 
+import os
+import pprint
+import threading
+import time
+from typing import Callable, Tuple, Union, NamedTuple, List
+from urllib.parse import urlparse
+
```
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/blockchain/eth/interfaces.py]: ✔️

File: nucypher/blockchain/eth/interfaces.py [hunk 8]
```diff
@@ -227,10 +239,11 @@ def __init__(self,
         self._provider = provider
         self._provider_process = provider_process
         self.w3 = NO_BLOCKCHAIN_CONNECTION
-        self.client = NO_BLOCKCHAIN_CONNECTION         # type: EthereumClient
+        self.client = NO_BLOCKCHAIN_CONNECTION
         self.transacting_power = READ_ONLY_INTERFACE
         self.is_light = light
         self.gas_strategy = gas_strategy or self.DEFAULT_GAS_STRATEGY
+        self.build_transaction_lock = threading.Lock()
```
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/blockchain/eth/interfaces.py]: ✔️

File: nucypher/blockchain/eth/sol/compile/aggregation.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+\|            # Anchored end of version definition | 
+\Z            # Anchor must be the end of the match
+""", re.VERBOSE)
+
+# simplified version of pattern to extract metadata hash from bytecode
+# see https://docs.soliditylang.org/en/latest/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
+METADATA_HASH_PATTERN: Pattern = re.compile(r"""
+a2
+64
+\w{8}       # 'i' 'p' 'f' 's'
```
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/blockchain/eth/sol/compile/aggregation.py]: Ok, ok, I was too lazy
✔️

File: nucypher/blockchain/eth/sol/compile/aggregation.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+58
+22
+\w{68}      # 34 bytes IPFS hash
+64
+\w{8}       # 's' 'o' 'l' 'c'
```
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/blockchain/eth/sol/compile/aggregation.py]: ✔️

File: tests/utils/blockchain.py (hunk 5 above)
Comment (vzotova) [reply tests/utils/blockchain.py]: ✔️

File: nucypher/blockchain/eth/interfaces.py [hunk 11]
```diff
@@ -945,7 +944,7 @@ def __get_contract_interface(self,
         """Retrieve compiled interface data from the cache and return web3 contract"""
         version, interface = self.find_raw_contract_data(contract_name, version)
         contract = self.client.w3.eth.contract(abi=interface['abi'],
-                                               bytecode=interface['bin'],
+                                               bytecode=interface['evm']['bytecode']['object'],
```
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/blockchain/eth/interfaces.py:945]: @KPrasch 👆

**Pull request merged**